### PR TITLE
Add automatic download retries with bounded backoff

### LINKED ISSUE #22: Add automatic download retries with bounded exponential backoff
## Motivation

A transient network failure immediately marks an item failed. It is retried only when a later job scan creates another run, although the model already exposes retry_count. Short outages therefore create avoidable failed runs.

## Proposed behavior

Retry transient failures within the same run using configurable maximum attempts and exponential backoff with jitter. Do not blindly retry permanent errors such as unsafe destination paths.

## Acceptance criteria

- Maximum attempts and delay bounds are configurable.
- retry_count increments and appears on the dashboard.
- Backoff does not block unrelated queued work.
- Transient and permanent errors have a documented classification.
- A successful retry completes the run normally.
- Exhaustion leaves a clear terminal failure.
- Shutdown cancels retry delays.
- Unit and SFTP tests cover success after retry and exhaustion.

### PR DESCRIPTION
**Pull request by IanWorley**
Closes #22

## Summary
- add validated retry attempt, delay, and jitter configuration with bounded exponential scheduling
- classify unsafe destination failures as permanent while retrying transient connection/transfer failures in the same run
- keep retry backoff durable and non-blocking, preserve dashboard retry counts, and surface clear terminal reasons
- document retry configuration and error classification
- cover eventual success, exhaustion, permanent failures, unrelated queue work, and shutdown cancellation

## Validation
- `dotnet build SporeSync.sln` — passed (existing warnings remain)
- `dotnet test SporeSync.sln` — passed: 270 tests
- Frontend checks not run; no UI files changed

### CODE REVIEW

File: SporeSync.Business/Worker/DownloadWorkerHostedService.cs [hunk 1]
```diff
@@ -487,6 +486,16 @@ await leafProgress.CompleteAsync(token => queueRepository.UpdateProgressAsync(ne
 
         if (groupFailed)
         {
+            if (groupPermanentFailure)
+            {
+                return await RecordPermanentFailureAsync(
```
Comment (IanWorley) [inline SporeSync.Business/Worker/DownloadWorkerHostedService.cs:491]: Important: This permanent group failure is not durable across scans. The group row becomes `failed`, but the next scan's `ChangeDetector.GroupMetadataChanged` returns true for any non-completed group, and unchanged failed leaves are not re-enqueued as changed leaves; they are carried forward/reused and `ProcessGroupAsync` reprocesses failed leaves on the group's next claim. For an unsafe destination inside a group, the scheduler will therefore retry the permanent failure on every later run, unlike single-file failed items which stay dead-lettered until remote change or manual retry. Fix by making unchanged failed/permanent groups follow the same dead-letter rule unless group or leaf metadata changed or manual retry is requested (for example, skip unchanged failed groups in change detection, or persist/check `handled_reason = permanent_error` when deciding whether to re-enqueue), and add a test covering an unchanged grouped permanent failure on the next scan.

**Review by IanWorley — COMMENTED** [review]
Round 1 review: I found one Important issue. CI is green and local backend build/test pass, but grouped permanent failures can be automatically retried by later scans, so the new permanent-error path is not durable for grouped downloads. I left the actionable detail inline.

File: SporeSync.Business/Worker/DownloadWorkerHostedService.cs (hunk 1 above)
Comment (IanWorley) [reply SporeSync.Business/Worker/DownloadWorkerHostedService.cs:491]: Fixed in `6f896fa`. `GroupMetadataChanged` now treats failed groups like other dead-lettered failed items: an unchanged failed group is not re-enqueued on a later scan, while group metadata changes or changed leaf metadata still revive it through the existing change paths. Added `DetectChanges_UnchangedPermanentlyFailedGroup_OnLaterScanStaysDeadLettered` to cover the unchanged permanent group + leaf rescan regression.

Validation:
- `dotnet build SporeSync.sln` — passed (existing warnings only)
- `dotnet test SporeSync.sln` — passed, 270/270 tests
- Frontend checks not run; no UI files changed.

**Review by IanWorley — COMMENTED** [review]
## Independent review — 6f896fa

### Important
The required `Unit Tests` check is failing for the PR merge result (runs [29219181797](https://github.com/IanWorley/SporeSync/actions/runs/29219181797) and [29219224474](https://github.com/IanWorley/SporeSync/actions/runs/29219224474)). `SftpSyncEndToEndTests.cs:221` and `:255` call `ISporeSyncRunRepository.CreateAsync`, while the current `main` base (`065ae2c`) renamed that API to `TryCreateAsync`. The head SHA alone still exposes `CreateAsync`, but the `pull_request` build compiles the merge result and fails CS1061 twice, so this PR cannot merge or run its test suite. Rebase/merge current `main` and update the two new retry E2E tests to the current run-creation API, handling its nullable return as appropriate.

I also re-checked the retry/worker/change-detection paths at the exact head; the previous failed-group dead-letter regression is addressed.

Validation evidence: inspected the exact `6f896fa12410a172bd645a878a36624dd34bd352` diff and current base API; both failed CI runs report the same CS1061 errors. Local .NET execution is unavailable in this Docker review environment (`dotnet` is not installed).

File: SporeSync.Business/Worker/DownloadWorkerHostedService.cs (hunk 1 above)
Comment (IanWorley) [reply SporeSync.Business/Worker/DownloadWorkerHostedService.cs:491]: Resolved in `05295cebc9741b7297e27b70d4298e545a42eaf9`. I rebased the PR onto current `main` (`065ae2c`), where run creation was renamed to nullable `TryCreateAsync`. The two retry SFTP E2E tests now call `TryCreateAsync` and use `Assert.IsType<SporeSyncRun>` before scanning, so a missing run fails the test explicitly rather than flowing as null. The branch was force-pushed with the rebased history.\n\nValidation: `git diff --check` passed. The assigned Docker backend does not include the .NET SDK (`dotnet: command not found`), so `dotnet build`/`dotnet test` could not run locally; CI has been triggered for the pushed SHA.

**Comment by IanWorley** [comment]
Resolved at 05295cebc9741b7297e27b70d4298e545a42eaf9: both retry E2E tests now call nullable `TryCreateAsync(job.Id)`, use `Assert.IsType<SporeSyncRun>(...)` to fail clearly when no run is created, and pass the proven non-null result to `ScanAsync`. The prior two SFTP `CreateAsync` CS1061 call sites are removed.\n\nThe Docker worker environment has no `dotnet`, so local build/test execution is unavailable. I inspected the current PR head and current interface signature; a fresh independent review is requested for this exact SHA.

**Review by IanWorley — COMMENTED** [review]
## Independent review

Reviewed exact head `05295cebc9741b7297e27b70d4298e545a42eaf9` against merge base `065ae2cb2082d4184992c58e3d6ce0f68b674986`.

No Blocker or Important findings in the retry/backoff changes. I checked the durable eligibility scheduling, retry-budget/dead-letter paths, permanent-path handling, grouped-download behavior, option validation, and tests. `git diff --check` is clean.

Required check status: **Unit Tests is failing** at the merge result because `RepositoryIntegrationTests.cs` calls nonexistent `SporeSyncRunRepository.CreateAsync`. This is a pre-existing base/main failure, not introduced by this PR: the PR head and its merge base use `TryCreateAsync`, while the current `main` workflow fails on the same `CreateAsync` error (run 29219941686). No PR-specific fix card created.

This is a COMMENT-only review; no approval or request-for-changes action was made.

**Review by IanWorley — COMMENTED** [review]
## Independent review — 05295ce

The two retry SFTP E2E tests now correctly assert a concrete `SporeSyncRun` from nullable `TryCreateAsync` before calling `ScanAsync`. I also verified the exact reviewed head has no remaining `.CreateAsync(...)` call outside documentation and `git diff --check 160ed757..05295ce` passes.

I am leaving this as a comment rather than approving because the required Unit Tests check is failing: https://github.com/IanWorley/SporeSync/actions/runs/29220117784/job/86723438635. Its log reports a stale-looking `CreateAsync` compile error in `RepositoryIntegrationTests`, but that call is not present in the reviewed 05295ce tree (the relevant test uses `TryCreateAsync`). Please rerun or diagnose the failed check against the exact head before merging.

Local .NET validation was unavailable in this reviewer environment (`dotnet: command not found`).

**Review by IanWorley — COMMENTED** [review]
## Review — compatibility fix at `05295cebc9741b7297e27b70d4298e545a42eaf9`

Reviewed only the two retry SFTP E2E call sites. No blocker or important finding in scope. `ISporeSyncRunRepository.TryCreateAsync` returns `Task<SporeSyncRun?>`, while `ISyncRunOrchestrator.ScanAsync` requires non-null `SporeSyncRun`; `Assert.IsType<SporeSyncRun>(...)` both checks the test precondition and yields a non-null value for the subsequent call.

Verification: the PR head matches the requested SHA. A fresh synthetic merge against the current base produced tree `a942088d20b45806a0d48b44f8c6b3fd7b3f8e68`, and the changed SFTP test file is byte-identical there. Docker review environment has no `dotnet`, so I could not execute the build/tests locally.

The failed CI run `29220117784` is not evidence against this exact head: its build log reports `RepositoryIntegrationTests.cs:1181` calling `CreateAsync`, whereas both the exact PR head and the fresh merge tree contain `TryCreateAsync` at that call site. A new CI run should be required for the actual head/merge result.

_No approval submitted; this is an independent COMMENT-only review._

**Review by IanWorley — COMMENTED** [review]
## Review — compatibility fix at `05295cebc9741b7297e27b70d4298e545a42eaf9`

Reviewed only the two retry SFTP E2E call sites. No blocker or important finding in scope. `ISporeSyncRunRepository.TryCreateAsync` returns `Task<SporeSyncRun?>`, while `ISyncRunOrchestrator.ScanAsync` requires non-null `SporeSyncRun`; `Assert.IsType<SporeSyncRun>(...)` checks the test precondition and yields a non-null value for the subsequent call.

Verification: the PR head matches the requested SHA. A fresh synthetic merge against the current base produced tree `a942088d20b45806a0d48b44f8c6b3fd7b3f8e68`, and the changed SFTP test file is identical there. This Docker review environment has no `dotnet`, so no local build/test ran.

The failed CI run `29220117784` is not evidence against this exact head: its build log reports `RepositoryIntegrationTests.cs:1181` calling `CreateAsync`, whereas both the exact PR head and fresh merge tree contain `TryCreateAsync` at that call site. Require a new CI run for the actual head/merge result.

_No approval submitted; this is an independent COMMENT-only review._

**Comment by IanWorley** [comment]
CI follow-up pushed at bf30fd5d37758c3147ad0e0a2df99a28e4dfaa97. The remaining RepositoryIntegrationTests atomic-create call now uses Assert.IsType<SporeSyncRun>(await TryCreateAsync(...)) before accessing run.Id, making the nullable API result explicit. Local dotnet build/test remain unavailable in this Docker backend; git diff --check and git show --check pass. New Unit Tests CI is running: https://github.com/IanWorley/SporeSync/actions/runs/29220289226. A fresh independent review has been requested for this exact SHA.

**Review by IanWorley — COMMENTED** [review]
## Independent review — bf30fd5

### Important
The required `Unit Tests` check fails for GitHub's generated merge result, so this PR cannot merge. The checkout was merge commit `a625d3c1db7d01cdc27caa8ebfa0919fbce8023a` (`bf30fd5` into base `160ed75`), and `SporeSync.Business.Tests/RepositoryIntegrationTests.cs:1181` still calls the removed `SporeSyncRunRepository.CreateAsync`, producing CS1061. The exact PR head has the compatibility change to `TryCreateAsync` at its corresponding test, but it is at a different location; the current base test remains unchanged in the merge result. Update the PR branch to resolve this merge-result build failure (for example, rebase/merge the current base and replace the stale call with the nullable-safe `TryCreateAsync` form), then rerun CI.

I also inspected the retry/backoff, permanent-failure, group, and change-detector changes at exact head `bf30fd5d37758c3147ad0e0a2df99a28e4dfaa97`; no additional Blocker or Important issue was identified. `git diff --check` passed for the merge result.

Validation evidence: [failed Unit Tests run](https://github.com/IanWorley/SporeSync/actions/runs/29220289226/job/86723912851) checks out `refs/remotes/pull/27/merge` at `a625d3c` and reports CS1061 at line 1181. Local .NET execution was unavailable because this Docker review environment does not have `dotnet`.

_COMMENT-only review; no approval or request-for-changes action was made._

**Comment by IanWorley** [comment]
Replying to the independent review: resolved in 99e42a75f5114cb33b196d6410aad77e85783a7c.

- Merged the current base `160ed75` into the existing PR branch so GitHub's merge-result tree contains the safe-configuration deletion tests.
- Replaced the stale `SporeSyncRunRepository.CreateAsync` call in `SporeSyncJobRepository_SafeDelete_BlocksActiveRun` with nullable-safe `Assert.IsType<SporeSyncRun>(await runRepository.TryCreateAsync(job.Id))`.
- Pushed the existing PR branch and verified CI: [Unit Tests passed](https://github.com/IanWorley/SporeSync/actions/runs/29220510090/job/86724522204) (including frontend lint/tests/build and .NET build/test); Build Artifact also passed.

Local `dotnet`/Docker are unavailable in this agent environment, so the verification above is GitHub CI. `git diff --check origin/main...HEAD` also passed locally.

Original review: https://github.com/IanWorley/SporeSync/pull/27#pullrequestreview-4681459949

**Review by IanWorley — COMMENTED** [review]
Review scope: historical commit bf30fd5d37758c3147ad0e0a2df99a28e4dfaa97.

The one-line change is correct: ISporeSyncRunRepository.TryCreateAsync explicitly returns SporeSyncRun?, and Assert.IsType<SporeSyncRun>(...) both documents the test precondition (a newly created job can queue a run) and narrows the nullable value before run.Id is used. No finding in this change.

CI run 29220289226 failed with CreateAsync at RepositoryIntegrationTests.cs(1181), but that line belongs to the then-current base/main safe-delete test in the PR merge checkout—not to bf30fd5, whose file blob has TryCreateAsync at its corresponding assertion. The branch follow-up 99e42a75 replaces that base-side CreateAsync use; its CI run 29220510090 has passed Build and Test.

**Review by IanWorley — COMMENTED** [review]
## Independent review — 99e42a75f511

Reviewed the exact head `99e42a75f5114cb33b196d6410aad77e85783a7c` against current base `160ed7573b6cb1ae77cc9817ad220253db3813d0`. I found no Blocker or Important issue in the retry/backoff, permanent-failure, grouped-download, change-detection, option-validation, and test changes. In particular, failed grouped state is now kept dead-lettered on an unchanged later scan, while a changed remote entry can re-enter the queue.

Evidence: `git diff --check origin/main...99e42a75f511` passed; a local three-way merge check found no conflict markers. GitHub currently reports `mergeable: MERGEABLE` / `mergeStateStatus: CLEAN`. Required checks are complete: `Unit Tests` and `Build Artifact` succeeded; the release/changelog jobs are skipped.

Local .NET build/test execution was not possible in this review container because `dotnet` is not installed (`dotnet --info` exits 127). This is a COMMENT-only review; no approval or request-for-changes action was made.

**Comment by IanWorley** [comment]
## ✅ Ready for Human Review

Reviewed SHA: `99e42a75f5114cb33b196d6410aad77e85783a7c`.

Fresh audit found no Blocker or Important findings. GitHub reports this exact head as `MERGEABLE` with `CLEAN` merge state; required checks are complete and green (`Unit Tests`, `Build Artifact`; release/changelog jobs are intentionally skipped). `git diff --check` and a synthetic three-way merge against current `main` are clean.

Local validation on an exact archived checkout: `dotnet build SporeSync.sln` passed; targeted retry/worker/change-detector/downloader/options tests passed (71/71). Full Testcontainers integration execution was not possible locally because Docker is unavailable; the required CI `Unit Tests` check is green.

COMMENT-only audit; no approval, request for changes, merge, or auto-merge action was made.

**Comment by IanWorley** [comment]
Conflict repaired by rebasing the existing PR branch onto current `main` and force-with-lease pushing `353e3c63e5163836142605231d9b68dcbe820bc2`. The resolution retains bounded retry/backoff coverage plus current host-key and atomic run-creation behavior.

Validation: `dotnet build SporeSync.sln` passed; 54 focused retry-related unit tests passed. Docker CLI is unavailable in this execution backend, so Testcontainers/full integration tests were not runnable here. GitHub reports this head as `MERGEABLE` (checks pending). Independent review card: `t_14f3f2e8`.

**Comment by IanWorley** [comment]
## Independent Review

Reviewed exact head `353e3c63e5163836142605231d9b68dcbe820bc2` against base `3e6cabc51cadbf308dea7d474f936210c8269256`. No Blocker or Important findings identified.

Evidence: `dotnet build SporeSync.sln --no-restore` passed; targeted retry/worker/change-detector/downloader/options tests passed (71/71). A full local Testcontainers suite was attempted but the execution environment cannot reach Docker; GitHub CI Unit Tests and Build Artifact completed successfully.

COMMENT-only independent review; no approval, request for changes, merge, or auto-merge action was made.

**Comment by IanWorley** [comment]
## ✅ Ready for Human Review

Verified final head SHA: `353e3c63e5163836142605231d9b68dcbe820bc2`.

Live GitHub reconciliation: PR is open, `MERGEABLE`, and `CLEAN`. Required CI checks completed successfully: `Unit Tests` and `Build Artifact`. Conditional release/changelog jobs are completed as skipped.

A fresh independent COMMENT-only review found no unresolved Blocker or Important findings. Local evidence: `dotnet build SporeSync.sln --no-restore` passed and the targeted retry/worker/change-detector/downloader/options suite passed (71/71). The full Testcontainers suite could not run locally because this environment cannot reach Docker; GitHub required Unit Tests CI is green.

No approval, request for changes, merge, or auto-merge action was made.
